### PR TITLE
Properly nest groups when using memberOf to detect group membership, …

### LINKED ISSUE #17759: 8.1 - LDAP: Having memberOf enabled breaks nested groups
### Steps to reproduce
1. Have an LDAP server, with memberOf enabled
2. Have owncloud configured for said LDAP server, with "Nested groups" enabled in the  "advanced settings"
3. Put LDAP user u1 in LDAP group g2
4. Put LDAP group g2 in LDAP group g1

### Expected behaviour
Owncloud should see the user u1 as a member of g1 (as well as g2)

### Actual behaviour
Owncloud only sees the user u1 as a member of g2

### Server configuration
**Operating system**: CentOS 7
**ownCloud version:** 8.1
**Are you using an external user-backend, if yes which one:** LDAP
**LDAP Server** OpenLDAP

### A bit of debugging...

090478a95e1adc904cd8971158c848b9c00374f6 added a speedup to the apps/user_ldap/group_ldap.php' getUserGroups() function, by using the memberOf attribute if possible (instead of checking all groups to see if the user's there, recursively if the nested groups option is enabled in the settings). However, memberOf itself is not recursive, thus nested groups are missed.

The code checks for memberOf support as well as the config option 'useMemberOfToDetectMembership', however, the said option is not exposed anywhere in the user interface...

I'd provide a PR but I'm unsure what would be the best solution between adding an option in the settings and not using memberOf if nested groups are enabled (or some third option I didn't think of)

@blizzz Thoughts, since you wrote that code?

### PR DESCRIPTION
**Pull request by GreenArchon**
…fixes #17759

### CODE REVIEW

**Comment by karlitschek** [comment]
@blizzz

**Comment by GreenArchon** [comment]
I hereby declare this diff MIT licensed

**Comment by LukasReschke** [comment]
@owncloud-bot This is okay to test.

File: apps/user_ldap/group_ldap.php [hunk 1]
```diff
@@ -182,6 +182,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
 	}
 
 	/**
+	 * @param string $dnGroup
+	 * @param array &$seen
+	 * @return array
+	 */
+	private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
```
Comment (LukasReschke) [inline apps/user_ldap/group_ldap.php]: array &$seen?

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please

File: apps/user_ldap/group_ldap.php [hunk 2]
```diff
@@ -182,6 +182,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+		if ($seen === null) {
+			$seen = array();
+		}
+		if (array_key_exists($DN, $seen)) {
+			// avoid loops
+			return array();
+		}
+		$seen[$DN] = 1;
+		$groups = $this->access->readAttribute($DN, 'memberOf');
+		if (is_array($groups)) {
```
Comment (MorrisJobke) [inline apps/user_ldap/group_ldap.php]: Please do it the other way around and negate this if statement. Then the indentation can be decreased.

Comment (MorrisJobke) [inline apps/user_ldap/group_ldap.php]:
```php
if($statement) {
  // do something here
} else {
  return [];
}
```

should get 

```php
if(!$statement) {
  return [];
}

// do something here
```

It's much more readable.

**Comment by GreenArchon** [comment]
Please re-review.

Also, keep in mind this comment from #17759 :
- "This is not the best code I've written, as there's now 3 different functions in group_ldap that basically do the same thing (find elements based on attributes of a certain type on a DN), but they don't exactly return the same thing, and I didn't want to get into that for a quick bugfix"

File: apps/user_ldap/group_ldap.php [hunk 3]
```diff
@@ -182,6 +182,36 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+		}
+		$seen[$DN] = 1;
+		$groups = $this->access->readAttribute($DN, 'memberOf');
+		if (!is_array($groups)) {
+			return array();
+		}
+		$groups = $this->access->groupsMatchFilter($groups);
+		$allGroups =  $groups;
+		foreach ($groups as $group) {
+			$nestedGroups = $this->access->connection->ldapNestedGroups;
```
Comment (blizzz) [inline apps/user_ldap/group_ldap.php]: I think we can check for this setting before entering foreach uselessly and just return what we found.

File: apps/user_ldap/group_ldap.php [hunk 4]
```diff
@@ -182,6 +182,36 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+			if (!empty($nestedGroups)) {
```
Comment (blizzz) [inline apps/user_ldap/group_ldap.php]: i don't trust this. please intval() and check for === 1.

File: apps/user_ldap/group_ldap.php [hunk 5]
```diff
@@ -377,14 +407,14 @@ public function getUserGroups($uid) {
 		if(intval($this->access->connection->hasMemberOfFilterSupport) === 1
 			&& intval($this->access->connection->useMemberOfToDetectMembership) === 1
 		) {
-			$groupDNs = $this->access->readAttribute($userDN, 'memberOf');
-
+			$groupDNs = $this->_getGroupDNsFromMemberOf($userDN);
+
```
Comment (blizzz) [inline apps/user_ldap/group_ldap.php]: while already at it, removing this tabs please.

**Comment by blizzz** [comment]
Looks good to me, especially with adjusted Tests! Only minor aspects to improve found. When they are done, :+1:  from me.

Then we need another tester / reviewer, maybe @tps800 @ryno83 @davitol

**Comment by blizzz** [comment]
@karlitschek  backport to 8.1 granted?

**Comment by karlitschek** [comment]
yes. please backport :-)

**Comment by GreenArchon** [comment]
@blizzz a55f233e9ffac7d492733f50a37343b4243898bf is good for you?

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
@blizzz any update ? Merging this is blocking the 8.1.2 backport https://github.com/owncloud/core/pull/18123

**Comment by MorrisJobke** [comment]
:+1:

**Comment by davitol** [comment]
:+1: Tested with "version":"8.2.0.4","versionstring":"8.2 pre alpha"

**Pull request merged**